aten/src/ATen/core/DistributionsHelper.h

```diff
@@ -0,0 +1,299 @@
+ */
+namespace at {
+
+// Using VectorType in Box-muller derived distributions to avoid
+// code duplication
+template <typename T>
+struct VectorType {  };
+
+template <> struct VectorType<float> {using type = FLOAT2;};
+template <> struct VectorType<at::Half> {using type = FLOAT2;};
+template <> struct VectorType<double> {using type = DOUBLE2;};
+
+template <typename T>
+using vect_type = typename VectorType<T>::type;
+
+// Constants for uniform distribution
+constexpr float POW_2_32_INV = 1.0f/std::numeric_limits<uint32_t>::max();
+constexpr double POW_2_64_INV = 1.0/std::numeric_limits<uint64_t>::max();
+
+/**
+ * Samples a uniform distribution in the range [0,1) of type T
+ */
+template <typename T>
+struct uniform_real_distribution {
+
+  C10_HOST_DEVICE inline uniform_real_distribution(T a_in, T b_in) {
+    #ifndef __HIP_PLATFORM_HCC__
+      #ifdef __CUDA_ARCH__
+        assert(a_in <= b_in);
+        assert(b_in-a_in <= std::numeric_limits<T>::max());
+      #else
+        AT_ASSERT(a_in <= b_in);
+        AT_ASSERT(b_in-a_in <= std::numeric_limits<T>::max());
+      #endif
+    #endif
+    a = a_in;
+    b = b_in;
+  }
+
+  template<typename U = T> 
+  C10_HOST inline typename std::enable_if<std::is_same<U, double>::value, double>::type
+  operator()(at::CPUGenerator* generator){
+    // See Note [Uniform Distribution Algorithm]
+    double x = generator->random64() * POW_2_64_INV;
+    if (x == 1.0) {
+      x = std::nextafter(1.0, 0.0);
+    }
+    return (x * (b - a) + a);
+
+  }
+
+  template<typename U = T> 
+  C10_HOST inline typename std::enable_if<std::is_same<U, float>::value
+                                          || std::is_same<U, at::Half>::value, float>::type
+  operator()(at::CPUGenerator* generator){
+    // See Note [Uniform Distribution Algorithm]
+    float x = generator->random() * POW_2_32_INV;
+    if (x == 1.0f) {
+      x = std::nextafter(1.0f, 0.0f);
+    }
+    return (x * (b - a) + a);
+  }
+
+  private:
+    T a;
+    T b; 
+};
+
+/**
+ * Samples a normal distribution using the Box-Muller method
+ * Takes mean and standard deviation as inputs
+ * Outputs two samples at a time
+ */
+template <typename T>
+struct normal_distribution {
+
+  C10_HOST_DEVICE inline normal_distribution(T mean_in, T stdv_in) {
+    #ifndef __HIP_PLATFORM_HCC__
+      #ifdef __CUDA_ARCH__
+        assert(stdv_in > 0);
+      #else
+        AT_ASSERT(stdv_in > 0);
+      #endif
+    #endif
+    mean = mean_in;
+    stdv = stdv_in;
+  }
+
+  C10_HOST inline vect_type<T> operator()(at::CPUGenerator* generator){
+    uniform_real_distribution<T> uniform(0.0, 1.0);
+    vect_type<T> result;
+    const T theta = static_cast<T>(2.0) * static_cast<T>(M_PI) * uniform(generator);
+    T u1 = uniform(generator);
+    // extra pre-caution to make sure log never gets zero
+    if (u1 == static_cast<T>(0.0)) {
+      u1 = std::numeric_limits<T>::min();
+    }
+    T r = ::sqrt(static_cast<T>(-2.0) * ::log(u1));
```

Comment:
It's not important now because CPU does not generate for halfs, but if you want to use this code on the gpu you'll need to you accumulateType to keep intermediates in float.

Reply:
I'll do it in the GPU PR